### DISCUSSION TITLE
Question: Should we allows multiple plugins uses same action keyword?

### DISCUSSION THREAD

**Discussion by taooceros** [Flow-Launcher/Flow.Launcher#510] [Ideas]
https://github.com/Flow-Launcher/Flow.Launcher/blob/69a5f291512edac4237269b1e047f89f15597f78/Flow.Launcher.Core/Plugin/PluginManager.cs#L122-L126

Based on this code, it seems that if action keyword conflict, only the last one will be kept.

**Comment by jjw24** [comment]
The current behaviour is warn and stop user from assigning duplicate action keywords.

Not sure how it behaves with downloaded plugin having duplicate action keywords, I think it might throw plugin loading error?

But current design is to not allow duplicates.

**Comment by taooceros** [comment]
> The current behaviour is warn and stop user from assigning duplicate action keywords.

That's true.
> Not sure how it behaves with downloaded plugin having duplicate action keywords, I think it might throw plugin loading error?

Not sure about that. I tried to modify the plugin.json file from plugin directory, and restart flow, but nothing happens. However, nothing in the code shows potential about plugin action keywords conflict will throw an error on loading, even though we do prevent users from assigning duplicate action keywords. Maybe we should add the check on it?

**Comment by jjw24** [comment]
What would actually happen if action keywords are duplicated in two plugins, you get two sets of results right, one from each plugin, but other than that any potential problems?

**Comment by taooceros** [comment]
> What would actually happen if action keywords are duplicated in two plugins, you get two sets of results right, one from each plugin, but other than that any potential problems?

Well, I think only one of the sets of results will be displayed, because the later one registering the action keyword will overwrite the earlier one.

**Comment by jjw24** [comment]
Oh ok, I think the behaviour then would be to load and disable the external plugin and output a message to let user know why it is disabled?

Or do we just want to automatically update it to a new keyword?

**Comment by taooceros** [comment]
Maybe we can allow duplicate key, so that we put the control to users about whether they want to search those two plugin with the same duplicate key.

> Or do we just want to automatically update it to a new keyword?

That seems also good, but we need to notify user when doing that.

**Reply by jjw24** [reply]
> Or do we just want to automatically update it to a new keyword?

relooking at this, probably not a good idea, this is going to cause unknown problems if keyword changes during runtime, even if we notify user i think it's just unnecessary logic to add.

**Comment by jjw24** [comment]
I am ok with allowing duplicate keys as long as there are no side effects

**Comment by jjw24** [comment]
@taooceros why is this moved to discussion as well, it's got pr linked

**Reply by taooceros** [reply]
It is moved because we haven't decided whether this should be implemented.

**Reply by jjw24** [reply]
ok cool

**Comment by Asahi21** [comment]
Hi, Ueli user here, I'm planning on moving over to Flow but not being able to use the same action keyword for many of my web searches is making it difficult, I use the same key for similar searches (like T for different torrent search engines) because it's easier to remember this way, Ueli also let's you select the order of priority when there are more than 1 plugin with the same action keyword (I believe this is already implemented), so in my opinion, it should be allowed.

**Reply by jjw24** [reply]
In addition to allowing duplicate action keywords, this feels like a similar use case as https://github.com/Flow-Launcher/Flow.Launcher/issues/438, allowing firing off multiple websites.

**Comment by jjw24** [comment]
thoughts @Flow-Launcher/team? i think dup action keyword is ok to go ahead.

**Reply by onesounds** [reply]
I vote for approval.

**Reply by Asahi21** [reply]
are there any news about this feature?

**Reply by jjw24** [reply]
This is a go ahead. Please allow sometime for us to revisit and develop this feature :)

**Reply by Asahi21** [reply]
Nice to hear! Thanks.
That's good news because I was going to start and try to edit the config files to enable multiple web searches using the same action keyword, as a workaround while I wait, the first time went well for a bit, but then broke when I tried to edit the web search plugin config through the GUI, so I have to start again.

I might just not try and wait depending on how long it takes to implement, I've already waited a year, what's some more lol

**Comment by Hernan-2017** [comment]
Hey, is there any news about this?

**Reply by jjw24** [reply]
Will go ahead with this feature but we will need to revisit the draft we made.

**Comment by onesounds** [comment]
@taooceros We need a work progress report here. You're working on this, right?

**Reply by jjw24** [reply]
PR in mention https://github.com/Flow-Launcher/Flow.Launcher/pull/227

**Comment by EnigmaSolved** [comment]
@jjw24, Was anything ever implemented for this (ie, for conducting a search via multiple plugins simultaneously)? 

A use case would be searching (via an Action Keyword) Google (via Web Searches), Browser History V2 (https://github.com/fahim-ahmed05/browser-history-v2), Browser Tabs (https://github.com/Flow-Launcher/Flow.Launcher.Plugin.BrowserTabs), and Browser Bookmarks all at the same time (ideally we'd be able to set priorities in relation to each other, eg, set Browser Tabs higher than Web Search). With auto-suggest turned on (where available) this would be similar to doing a search from the browser address bar. 😊

**Reply by jjw24** [reply]
Hey there, this has been added in the prerelease version already.

**Reply by EnigmaSolved** [reply]
Oh wow, good to hear -- thanks for the reply, @jjw24!
